2024 03 29 send to replicas errors

andrewbaptist opened the pull request.

Commit:
kvclient: extract transport construction

Extract the creation of the transport into a seperate method. Previously
this was done inside sendToReplicas and only needed once. Since the
creation of the transport has become more complicated, extracting this
out simplifies the flow of sendToReplicas and allows calling this method
from additional places.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/121368)
<!-- Reviewable:end -->

Commit:
kvcoord: retry on a compatible NLHE

Previously if we received a NLHE with a compatible transport, we would
error out of sendToReplicas, evict from our cache and expect to get
called again with a new routing from the cache. This was unnecessary
since we had all the information necessary to continue already.

Epic: none

Release note: None

Commit:
kvclient: don't drop ambiguous errors on incompatible transport

Previously in the case of a NLHE with an incompatible transport we would
exit the retry loop immediately and return the error without checking if
there were any ambiguous or replica unavailable errors. This had a risk
of losing an ambiguous error if the retry lt ater succeeded and returning
an incorrect result to the caller.

Epic: none

Release note: None

Commit:
kvclient: add comment describing the return values

sendToReplicas has a fairly complex contract specfically around the
error cases. There have been a number of subtle bugs related to how it
is handled. The goal of this PR is to document the current behavior. In
future PRs the behavior and this comment can be simplified.

Epic: none

Release note: None

Commit:
kvclient: remove routeToLeaseholder variable

FIXME: This causes a data race. We can't modify the passed in BR safely

Previously we extracted a routeToLeaseholder variable which was intended
to control whether we sent the request to the leaseholder or a follower
first. It is better to modify the BatchRequest to correctly track
whether we are routing to the leaseholder. This has the side effect of
allowing a higher level retry to bypass an unnecessary attempt to hit
the follower first.

Epic: none

Release note: None

Commit:
kvclient: better testing of NLHE

Internally we handle NLHE errors differently depending on whether they
contain a new descriptor or the old one. This test validates the error
conditions in both cases.

Epic: none

Release note: None
